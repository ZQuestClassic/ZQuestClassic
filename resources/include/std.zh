// std.zh
// 28th December, 2020 for ZC 2.55 Alpha 86
// Standard header for ZScript.
// Please don't edit this file. Instead, create a new header file for your
// quests' scripts,  add constants and procedures to that file,
// and #include it separately.

//Options
#option BINARY_32BIT off
	//Would break a number of things if set on
#option SHORT_CIRCUIT on
	//Should work for everything in std_zh, and just speed it up

#includepath "include"

//Version Metadata
#include "std_zh/std_meta.zh"

//User Configuration Files
#include "std_zh/std_user_defs.zh" //Typedefs: Always first.
#include "std_zh/std.cfg"

//Main Files
#include "std_zh/limits.zh"
#include "std_zh/std_constants.zh"
#include "std_zh/ffrules.zh"
#include "std_zh/std_functions.zh"
#include "std_zh/math.zh"
#include "std_zh/weapon.zh"
#include "std_zh/std_sideview.zh" //Sideview and platforming

//ZScript string handling functions, akin to those found in C
//#include "string.zh"
#include "std_zh/string/string_constants.zh" //Now #includeed from here. This prevents conflicts for those people
#include "std_zh/string/string_functions.zh" //who would #include string.zh from a secondary source, as that file is now 
			     //empty, and the string.zh components load from here. 

//Keyboard handling for 2.54 and above
#include "std_zh/std_keyboard.zh"

//Locale
#include "std_zh/locale/std_EnglishIntl.zh" //International English Localisation

//Legacy Wrappers
#include "std_zh/std_legacy.zh"