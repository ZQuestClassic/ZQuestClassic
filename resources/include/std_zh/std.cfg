//Settings

const int STD_FORCE_OLD_DISTANCE = 1;	//If enabled, Distance() will use Sqrt() instead of SafeSqrt().
					//! If you use this, disable the quest rule 'Log Script Errors to Allegro.log
					//! before distributing your quest!
					
const int STD_DISTANCE_USE_SAFESQRT = 0; 	//If enebaled, Distance will use SafeSqrt(). Otherwise, it will return a factor, 
					
const int STD_EMULATE_ENGINE_COLLISION = 0; // If enabled, Collision hitboxs will more precisely emulate the
					    // internal collision. This differs to how scripted collision
					    // has always worked.
					    

const int STD_CENTERXY_USES_HITBOXES = 0; // If enabled, CenterX() and CenterY() will use object hitboxes
					  // and offsets by default, rather than sprite sizes when calculating
					  // the centre of objects.
					  

const int STD_NEVER_USE_SAFESQRT = 1; //Enable to force any calls to SafeSqrt() to use Sqrt() instead.

const int MAX_HIGHEST_LEVEL_ITEM_CHECK = 255; //The highest item ID checked by GetHighestLevelItem(). 

const int REVISED_ROTATEPOINT_X_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointX()
const int REVISED_ROTATEPOINT_Y_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointY()


const int MAX_NOACTION_INPUT = 14; //Checks A, B, L, R, U/D/L/R, Ex1, Ex2, Ex3, Ex4
	//Use CB_MAX to force the analogue stick false. 
const int NOACTION_SKIP_MAP = 0; //If enabled, NoAction will not disable pressing the Map button
const int NOACTION_SKIP_START = 0; //If enabled, NoAction will not disable pressing the Start button

const int OLD_ONSIDEPLATFORM = 0; //Change to 1 to use old OnSidePlatform calls.

const int OLD_RAND_FUNCS = 0; //Change to 1 to use old Rand() logic
