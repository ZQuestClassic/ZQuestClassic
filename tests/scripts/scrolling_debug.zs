#include "std.zh"

dmapdata script ScrollingDebug
{
    void run()
    {
        while (true)
		{
			printf("OX/OY %d %d\n", Game->Scrolling[SCROLL_OX], Game->Scrolling[SCROLL_OY]);
			Screen->DrawString(7, 0, 0, FONT_Z3SMALL, 0x01, 0x0F, TF_NORMAL, "OX/OY", 128);
			Screen->DrawInteger(7, 40, 0, FONT_Z3SMALL, 0x01, 0x0F, -1, -1, Game->Scrolling[SCROLL_OX], 0, 128);
			Screen->DrawInteger(7, 120, 0, FONT_Z3SMALL, 0x01, 0x0F, -1, -1, Game->Scrolling[SCROLL_OY], 0, 128);

			printf("NX/NY %d %d\n", Game->Scrolling[SCROLL_NX], Game->Scrolling[SCROLL_NY]);
			Screen->DrawString(7, 0, 8, FONT_Z3SMALL, 0x01, 0x0F, TF_NORMAL, "NX/NY", 128);
			Screen->DrawInteger(7, 40, 8, FONT_Z3SMALL, 0x01, 0x0F, -1, -1, Game->Scrolling[SCROLL_NX], 0, 128);
			Screen->DrawInteger(7, 120, 8, FONT_Z3SMALL, 0x01, 0x0F, -1, -1, Game->Scrolling[SCROLL_NY], 0, 128);

			if (Game->Scrolling[SCROLL_DIR] != -1)
			{
				// This draws where link is.
				int x = Link->X + Game->Scrolling[SCROLL_NX];
				int y = Link->Y + Game->Scrolling[SCROLL_NY];
				Screen->DrawCombo(3, x, y,
					20,
					1,
					1,
					3,
					-1, -1, 0, 0, 0, 0, 0, true, OP_OPAQUE);

				// This draws in the middle of the old screen.
				x = 256/2 + Game->Scrolling[SCROLL_OX];
				y = 176/2 + Game->Scrolling[SCROLL_OY];
				Screen->DrawCombo(3, x, y,
					21,
					1,
					1,
					3,
					-1, -1, 0, 0, 0, 0, 0, true, OP_OPAQUE);
				
				// This draws in the middle of the new screen.
				x = 256/2 + Game->Scrolling[SCROLL_NX];
				y = 176/2 + Game->Scrolling[SCROLL_NY];
				Screen->DrawCombo(3, x, y,
					44,
					1,
					1,
					3,
					-1, -1, 0, 0, 0, 0, 0, true, OP_OPAQUE);
			}

			Waitframe();
		}
    }
}
