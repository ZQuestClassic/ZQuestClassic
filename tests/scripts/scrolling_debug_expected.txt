PUSHR REFDMAPDATA
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
SETR d4,SP
ALLOCATEMEMV d2,13
STORED d2,5
WRITEPODSTRING d2,"OX/OY %d %d\n"
LOADD d2,5
PUSHR d2
SETV d2,3
PUSHR d2
POP d0
PUSHVARGR GAMESCROLLING
SETV d2,4
PUSHR d2
POP d0
PUSHVARGR GAMESCROLLING
PRINTFVARG
POP d5
LOADD d3,5
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,6
STORED d2,4
WRITEPODSTRING d2,"OX/OY"
PUSHV 7
PUSHV 0
PUSHV 0
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV 0
LOADD d2,4
PUSHR d2
PUSHV 128
DRAWSTRING
POPARGS d5,0.0009
LOADD d3,4
DEALLOCATEMEMR d3
PUSHV 7
PUSHV 40
PUSHV 0
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV -1
PUSHV -1
SETV d2,3
PUSHR d2
POP d0
PUSHR GAMESCROLLING
PUSHV 0
PUSHV 128
DRAWINT
POPARGS d5,0.0011
PUSHV 7
PUSHV 120
PUSHV 0
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV -1
PUSHV -1
SETV d2,4
PUSHR d2
POP d0
PUSHR GAMESCROLLING
PUSHV 0
PUSHV 128
DRAWINT
POPARGS d5,0.0011
ALLOCATEMEMV d2,13
STORED d2,3
WRITEPODSTRING d2,"NX/NY %d %d\n"
LOADD d2,3
PUSHR d2
SETV d2,1
PUSHR d2
POP d0
PUSHVARGR GAMESCROLLING
SETV d2,2
PUSHR d2
POP d0
PUSHVARGR GAMESCROLLING
PRINTFVARG
POP d5
LOADD d3,3
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,6
STORED d2,2
WRITEPODSTRING d2,"NX/NY"
PUSHV 7
PUSHV 0
PUSHV 8
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV 0
LOADD d2,2
PUSHR d2
PUSHV 128
DRAWSTRING
POPARGS d5,0.0009
LOADD d3,2
DEALLOCATEMEMR d3
PUSHV 7
PUSHV 40
PUSHV 8
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV -1
PUSHV -1
SETV d2,1
PUSHR d2
POP d0
PUSHR GAMESCROLLING
PUSHV 0
PUSHV 128
DRAWINT
POPARGS d5,0.0011
PUSHV 7
PUSHV 120
PUSHV 8
PUSHV 2
PUSHV 1
PUSHV 15
PUSHV -1
PUSHV -1
SETV d2,2
PUSHR d2
POP d0
PUSHR GAMESCROLLING
PUSHV 0
PUSHV 128
DRAWINT
POPARGS d5,0.0011
SETV d2,0
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
COMPAREV d2,-1
SETFALSEI d2
COMPAREV d2,0
GOTOTRUE 251
SETR d2,LINKX
PUSHR d2
SETV d2,1
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
POP d3
ADDR d2,d3
STORED d2,1
SETR d2,LINKY
PUSHR d2
SETV d2,2
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
POP d3
ADDR d2,d3
STORED d2,0
PUSHV 3
LOADD d2,1
PUSHR d2
LOADD d2,0
PUSHR d2
PUSHV 20
PUSHV 1
PUSHV 1
PUSHV 3
PUSHV -1
PUSHV -1
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 1
PUSHV 128
DRAWCOMBO
POPARGS d5,0.0016
SETV d2,3
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
ADDV d2,128
STORED d2,1
SETV d2,4
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
ADDV d2,88
STORED d2,0
PUSHV 3
LOADD d2,1
PUSHR d2
LOADD d2,0
PUSHR d2
PUSHV 21
PUSHV 1
PUSHV 1
PUSHV 3
PUSHV -1
PUSHV -1
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 1
PUSHV 128
DRAWCOMBO
POPARGS d5,0.0016
SETV d2,1
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
ADDV d2,128
STORED d2,1
SETV d2,2
PUSHR d2
POP d0
SETR d2,GAMESCROLLING
ADDV d2,88
STORED d2,0
PUSHV 3
LOADD d2,1
PUSHR d2
LOADD d2,0
PUSHR d2
PUSHV 44
PUSHV 1
PUSHV 1
PUSHV 3
PUSHV -1
PUSHV -1
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 0
PUSHV 1
PUSHV 128
DRAWCOMBO
POPARGS d5,0.0016
WAITFRAME
GOTO 9
POPARGS d5,0.0007
QUIT


ALLOCATEGMEMV d2,4096
SETR gd1,d2
QUIT